diff --git a/src/shell.js b/src/shell.js
index c1e0998c5..a36e268b4 100644
--- a/src/shell.js
+++ b/src/shell.js
@@ -145,7 +145,7 @@ var quit_ = (status, toThrow) => {
 #if SHARED_MEMORY && !MODULARIZE
 // In MODULARIZE mode _scriptName needs to be captured already at the very top of the page immediately when the page is parsed, so it is generated there
 // before the page load. In non-MODULARIZE modes generate it here.
-var _scriptName = (typeof document != 'undefined') ? document.currentScript?.src : undefined;
+var _scriptName = (typeof document != 'undefined' && document.currentScript?.tagName.toUpperCase() === 'SCRIPT') ? document.currentScript.src : undefined;
 
 #if ENVIRONMENT_MAY_BE_NODE
 if (ENVIRONMENT_IS_NODE) {
@@ -341,7 +341,7 @@ if (ENVIRONMENT_IS_SHELL) {
 if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
   if (ENVIRONMENT_IS_WORKER) { // Check worker, not web, since window could be polyfilled
     scriptDirectory = self.location.href;
-  } else if (typeof document != 'undefined' && document.currentScript) { // web
+  } else if (typeof document != 'undefined' && document.currentScript?.tagName.toUpperCase() === 'SCRIPT') { // web
     scriptDirectory = document.currentScript.src;
   }
 #if MODULARIZE
diff --git a/src/shell_minimal.js b/src/shell_minimal.js
index 60c708960..bb0d70dd2 100644
--- a/src/shell_minimal.js
+++ b/src/shell_minimal.js
@@ -134,7 +134,7 @@ var ENVIRONMENT_IS_PTHREAD = ENVIRONMENT_IS_WORKER && self.name?.startsWith('em-
 #if !MODULARIZE
 // In MODULARIZE mode _scriptName needs to be captured already at the very top of the page immediately when the page is parsed, so it is generated there
 // before the page load. In non-MODULARIZE modes generate it here.
-var _scriptName = (typeof document != 'undefined') ? document.currentScript?.src : undefined;
+var _scriptName = (typeof document != 'undefined' && document.currentScript?.tagName.toUpperCase() === 'SCRIPT') ? document.currentScript.src : undefined;
 #endif
 
 #if ENVIRONMENT_MAY_BE_NODE
diff --git a/tools/link.py b/tools/link.py
index ea7938f11..9d7ff7bfe 100644
--- a/tools/link.py
+++ b/tools/link.py
@@ -2427,7 +2427,7 @@ export default %(maybe_async)s function init(moduleArg = {}) {
     if settings.EXPORT_ES6:
       script_url = 'import.meta.url'
     else:
-      script_url = "typeof document != 'undefined' ? document.currentScript?.src : undefined"
+      script_url = "(typeof document != 'undefined' && document.currentScript?.tagName.toUpperCase() === 'SCRIPT') ? document.currentScript.src : undefined"
       if settings.ENVIRONMENT_MAY_BE_NODE:
         script_url_node = "if (typeof __filename != 'undefined') _scriptName = _scriptName || __filename;"
     if settings.MODULARIZE == 'instance':
